devnet testing sdk prototype

sigma opened the pull request:
**Description**

This change introduces:
- a basic DSL for expressing high-level system tests
- an implementation of a system, backed by kurtosis devnet provisioner
- some testing helpers to make for a better dev UX
- a basic Contract API for demonstration purpose

<!--
A clear and concise description of the features you're adding in this pull request.
-->

**Tests**

<!--
Please describe any tests you've added. If you've added no tests, or left important behavior untested, please explain why not.
-->

**Additional context**

<!--
Add any other context about the problem you're solving.
-->

**Metadata**

<!-- 
Include a link to any github issues that this may close in the following form:
- Fixes #[Link to Issue]
-->

sigma committed:
feat(devnet-sdk): expose minimal test interface

sigma committed:
feat(devnet-sdk): basic infrastructure for mini test

Co-authored-by: Jan Nanista <jan@oplabs.co>

scharissis committed:
feat(devnet-sdk): implemented system

sigma committed:
test(interop): new mini test to validate framework

sigma committed:
test(interop): improve mini test

sigma committed:
fix(devnet-sdk): improve cnx management

sigma committed:
fix(devnet-sdk): use big.Int

sigma committed:
test(interop): add constraints

sigma committed:
fix(devnet-sdk): fix tests

sigma committed:
fix(devnet-testing): use pointer receivers consistently

sigma committed:
fix(devnet-testing): remove unnecessary ContractAddress

sigma committed:
fix(devnet-testing): remove unnecessary NewBalanceFromFloat

sigma committed:
fix: cosmetic change to interop smoke test

sigma committed:
feat(devnet-sdk): proper interface-backed implementation

sigma committed:
feat(devnet-testing): testing the tests

sigma committed:
feat(devnet-testing): interface cleanup

scharissis commented on devnet-sdk/contracts/registry/client/client.go:
```diff
@@ -0,0 +1,28 @@
+	Client *ethclient.Client
+}
+
+func (r *ClientRegistry) SuperchainWETH(address types.Address) (interfaces.SuperchainWETH, error) {
```
Should this implement the `Registry` interface?

sigma replied:
it implements ContractsRegistry, as stated above

"Registry" was a left-over from an earlier refactoring

scharissis commented on devnet-sdk/contracts/registry/client/superchainweth.go:
```diff
@@ -0,0 +1,36 @@
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+type superchainWETHBinding struct {
```
nit/suggestion: add an interface satisfaction check to show that this implements `SuperchainWETH interface`.

sigma replied:
done

scharissis commented on devnet-sdk/contracts/registry/empty/empty.go:
```diff
@@ -0,0 +1,16 @@
+// EmptyRegistry represents a registry that returns not found errors for all contract accesses
+type EmptyRegistry struct{}
+
+func (r *EmptyRegistry) SuperchainWETH(address types.Address) (interfaces.SuperchainWETH, error) {
```
Again, should this implement the `Registry` interface?

sigma replied:
same here. ContractsRegistry is the interface

scharissis commented on devnet-sdk/interfaces/registry.go:
```diff
@@ -0,0 +1,27 @@
+package interfaces
```
Is this package intended for all interfaces?

SuperchainWETH already exists in `contracts/contracts.go`.

sigma replied:
not in the latest version of the branch!
aka: you're right, but I did the change already :)

sigma committed:
fix: remove obsolete Registry definition

scharissis commented on devnet-sdk/system/chain.go:
```diff
@@ -0,0 +1,137 @@
+	return c.rpcUrl
+}
+
+func (c *chain) Wallet(ctx context.Context, constraints ...constraints.WalletConstraint) (types.Wallet, error) {
```
```suggestion
// Wallet returns the first user which meets all provided constraints, or an error.
func (c *chain) Wallet(ctx context.Context, constraints ...constraints.WalletConstraint) (types.Wallet, error) {
```

sigma replied:
done

scharissis commented on devnet-sdk/system/chain.go:
```diff
@@ -0,0 +1,137 @@
+}
+
+func chainFromDescriptor(d *descriptors.Chain) Chain {
+	firstNodeRPC := d.Nodes[0].Services["el"].Endpoints["rpc"]
```
IIRC, I wrote this. Still, couldn't this panic? :)

sigma replied:
yeah I guess it could. Although honestly, in a test scenario that's probably not the worst thing.
Let me add a TODO to make things a bit better

scharissis commented on devnet-sdk/system/interfaces.go:
```diff
@@ -0,0 +1,35 @@
+	"github.com/ethereum-optimism/optimism/devnet-sdk/types"
+)
+
+// System represents a complete Optimism system with L1 and L2 chains
+type System interface {
+	Identifier() string
+	L1() Chain
+	L2(uint64) Chain
+}
```
L2 Chains (plural)? The definition would imply singular.

sigma replied:
well, it's the *single* chain with the provided ID.

Note that this is already wrong because IDs don't always fit in a uint64.
Let me add a TODO, as it kinda works for now

scharissis commented on devnet-sdk/system/interfaces.go:
```diff
@@ -0,0 +1,35 @@
+	InteropSet() InteropSet
+}
+
+// InteropSet provides access to L2 chains in an interop environment
```
Again, one or many L2's?

sigma replied:
Well, same argument: each L2 is uniquely identified by its chain ID

scharissis commented on devnet-sdk/system/wallet.go:
```diff
@@ -0,0 +1,199 @@
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+// internalChain provides access to internal chain functionality
+type internalChain interface {
```
It's not clear to me; why not just expose it then?

sigma replied:
well, mostly because it would expose the implementation detail that a ethclient.Client is involved, which we explicitly do not want (as we need to be able to fake a chain, that is not backed by an actual client connection).
But we could also say that we just don't really need it. I used it solely as a forcing function to not have downstream code depend on additional implementation details of the actual chain object defined in chain.go

scharissis commented on devnet-sdk/system/wallet.go:
```diff
@@ -0,0 +1,199 @@
+		return nil, fmt.Errorf("failed to get gas price: %w", err)
+	}
+
+	gasLimit := uint64(210000) // 10x Standard ETH transfer gas limit
```
Should this be a variable, or a package constant?

sigma replied:
fair question. Right now that is a hack to make things work.
Ideally this should be dynamically computed.

scharissis commented on devnet-sdk/testing/systest/systest.go:
```diff
@@ -0,0 +1,49 @@
+	"github.com/ethereum-optimism/optimism/devnet-sdk/system"
+)
+
+type Validator func(t T, sys system.System) (context.Context, error)
```
What does a validator do? Check if a System is in a valid state (to facilitate skips)? Or is it the actual test to be conducted?

sigma replied:
That's a test precondition validator, definitely not the test itself.
Let me clarify the name.

scharissis commented on devnet-sdk/types/balance.go:
```diff
@@ -0,0 +1,75 @@
+	"math/big"
+)
+
+type Balance struct {
```
Tests for these Balance func's would be great.

sigma replied:
fair. added some

sigma committed:
fix: address review comments

- add explicit interface checks (and remove superfluous interface)
- add tests for Balance type

scharissis reviewed (approved):
Looks Great To Me! :)
